src/api_manager/context/request_context_test.cc

```diff
@@ -0,0 +1,393 @@
+  std::function<void()> continuation_;
+};
+
+class MockApiManagerEnvironment : public ApiManagerEnvInterface {
+ public:
+  virtual ~MockApiManagerEnvironment() {}
+
+  void Log(LogLevel level, const char *message) override {
+    // std::cout << __FILE__ << ":" << __LINE__ << " " << message << std::endl;
+  }
+  std::unique_ptr<PeriodicTimer> StartPeriodicTimer(
+      std::chrono::milliseconds interval, std::function<void()> continuation) {
+    return std::unique_ptr<PeriodicTimer>(new MockPeriodicTimer(continuation));
+  }
+
+  void RunHTTPRequest(std::unique_ptr<HTTPRequest> request) override {}
+
+  virtual void RunGRPCRequest(std::unique_ptr<GRPCRequest> request) override {}
+};
+
+class MockRequest : public Request {
+ public:
+  MockRequest(const std::string &client_ip,
+              const std::map<std::string, std::string> &header)
+      : client_ip_(client_ip), header_(header) {}
+
+  virtual ~MockRequest() {}
+
+  bool FindHeader(const std::string &name, std::string *header) override {
+    auto it = header_.find(name);
+    if (it != header_.end()) {
+      header->assign(it->second);
+      return true;
+    }
+
+    return false;
+  }
+
+  std::string GetClientIP() { return client_ip_; }
+
+  std::string GetRequestHTTPMethod() override { return "GET"; }
+
+  std::string GetQueryParameters() override { return ""; }
+
+  std::string GetUnparsedRequestPath() override { return "/echo"; }
+
+  bool FindQuery(const std::string &name, std::string *query) override {
+    return false;
+  }
+
+  MOCK_METHOD2(AddHeaderToBackend,
+               utils::Status(const std::string &, const std::string &));
+  MOCK_METHOD1(SetAuthToken, void(const std::string &));
+  MOCK_METHOD0(GetFrontendProtocol,
+               ::google::api_manager::protocol::Protocol());
+  MOCK_METHOD0(GetBackendProtocol, ::google::api_manager::protocol::Protocol());
+  MOCK_METHOD0(GetRequestPath, std::string());
+  MOCK_METHOD0(GetInsecureCallerID, std::string());
+  MOCK_METHOD0(GetRequestHeaders, std::multimap<std::string, std::string> *());
+  MOCK_METHOD0(GetGrpcRequestBytes, int64_t());
+  MOCK_METHOD0(GetGrpcResponseBytes, int64_t());
+  MOCK_METHOD0(GetGrpcRequestMessageCounts, int64_t());
+  MOCK_METHOD0(GetGrpcResponseMessageCounts, int64_t());
+
+ private:
+  const std::string client_ip_;
+  const std::map<std::string, std::string> header_;
+};
+}
+
+class RequestContextTest : public ::testing::Test {
+ protected:
+  RequestContextTest() : callback_run_count_(0) {}
+
+  void SetUp() {
+    callback_run_count_ = 0;
+    call_history_.clear();
+  }
+
+ protected:
+  std::vector<std::string> call_history_;
+  int callback_run_count_;
+};
+
+TEST_F(RequestContextTest, ClientIPAddressNoOverrideTest) {
+  std::unique_ptr<ApiManagerEnvInterface> env(new MockApiManagerEnvironment());
+
+  std::shared_ptr<context::GlobalContext> global_context(
+      new context::GlobalContext(
+          std::move(env), std::string(kServerConfigWithoutClientIPExperiment)));
```

Comment:
the cont server_config string can be defined in this test function so it is easier to read 
instead of moving back and forc

Reply:
Relocated server config into the each function